### PR TITLE
Datepicker should use Date as state model, (local-now) for determining today

### PR DESCRIPTION
**Pull request by osbert**
When determining the value of *today* for highlighting purposes, we should use (local-now) so that it matches the end user's current date as opposed to the current UTC date.

The underlying model used to store the current value of the Datepicker should be a Date object since the widget does not actually capture both a Date and a Time. This then allows the component user to further manipulate as needed without worrying about time zone conversions or other issues.

### CODE REVIEW

**Comment by hipitihop** [comment]
@osbert thanks for the PR. I have had a look at this and I still have some reservations, although I'm more then happy to admit I'm missing something subtle. I'm still doing some more experiments but I wanted to acknowledge your submission and give some preliminary feedback.
-  While I agree with your argument that the widget only deals with dates, not time, even goog.date.Date (as do its counterparts DateTime & UtcDateTime) is based on milliseconds since epoch and it simply forces the time to zero, but still seems to have a notion of timezone. e.g. if you do `(.getTimezoneOffset (js/goog.date.Date.))` what do you get ? I get -600 I am based in Queensland, Australia so GMT+1000 (AEST)
- I have experimented with (now) used in [datepicker.cljs line 201](https://github.com/Day8/re-com/blob/master/src/re_com/datepicker.cljs#L201) for :today which in turn is defined in [cljs-time.core L247-L250](https://github.com/andrewmcveigh/cljs-time/blob/master/src/cljs_time/core.cljs#L247-L250) in terms of (goog.date.UtcDateTime.) My tests so far suggest that although this is a goog.date.UtcDateTime that the date/time is indeed based on users local date & time (although I have only tested this on my Ubuntu/Linux machine) If this is the case and since all datepicker underlying dates model, minimum, maximum & today are goog.date.UtcDateTime instances, I'm struggling to see how `today` could be out. Perhaps some sample code that demonstrates the issue might help me have an aha moment.
- I understand that depending on context, consumers may optionally want to deal with a localised date instance after selection, I would suggest to convert any selected date back to a local one e.g. in your registered handler:<br>
`:on-change (fn [goog-utc] (cljs-time.coerce/to-local-date goog-utc))`
- We have found this area of timezone based dates is where dragons live, particularly if you try and serialise and transfer dates across timezones or simply to a server which may be in a different timezone, if you are not careful a date created in one is not the same elsewhere and you may want it to be a common currency. Don't even get me started if you further consider daylight saving DST which is not represented in the timezone offsets.

**Comment by osbert** [comment]
I appreciate the timely response. I will try to explain the exact problem a little bit better. Note that there is technically nothing *wrong* with the code nor the reasoning which is why I understand your concern completely. However, as a result it is somewhat difficult for me to illustrate using a fully automated test but here's my best effort given time available. I'll continue thinking about a good way to express this in a test but hopefully this helps.

```clojure
(ns re-com.test.datepicker-test
  (:require-macros [cemerick.cljs.test :refer (is are deftest with-test run-tests testing test-var)])
  (:require [cemerick.cljs.test]
            [reagent.core :as reagent]
            [re-com.datepicker :as datepicker]
            [cljs-time.core :as t]
            [cljs-time.coerce :as tc]
            [cljs-time.format :as tf]))

(deftest test-datepicker-today-highlight
  ;; The highlight is controlled by whether or not (=date date today)
  ;; return true, in which case it adds the style "today".
  ;;
  ;; In this check, date refers to the calendar date being rendered,
  ;; which is generated by iterating the underlying model in
  ;; increments of days, and of type goog.date.UtcDateTime.
  ;;
  ;; To populate "today", we call (now) in (datepicker/configure)
  ;; which also returns UtcDateTime. This is all exactly as you
  ;; stated.
  ;;
  ;; Let's assume that the user is in Australia (+10 offset).
  ;; This means that when it is currently local time 0700, in GMT it
  ;; is 2100 the previous day.
  ;;
  ;; To give a full concrete example, this means that at local time
  ;; 2015-05-11 0700 (+10), we'll have the following:
  ;; 
  ;;   today => (now) => 2015-05-10 21:00 (UTC)
  ;; 
  ;; Since we're looking for (date= today date), this predicate will
  ;; return true when date is 2015-05-10 00:00 (UTC), and thus
  ;; highlight 2015-05-10 even though the user local time is currently
  ;; 2015-05-11.
  ;;
  (let [today (goog.date.UtcDateTime. 2015 5 10 21 0)] ;; 2015-05-10 21:00 (UTC)

    ;; This illustrates that it will highlight 2015-05-10. For all
    ;; intents and purposes, this is actually correct behavior. These
    ;; invariants should actually always be true in this code.
    (is (datepicker/=date today (goog.date.UtcDateTime. 2015 5 10)))
    (is (not (datepicker/=date today (goog.date.UtcDateTime. 2015 5 11))))
    
    ;; However, the actual problem is that for the user, highlighting
    ;; 2015-05-10 is incorrect as it is currently 2015-05-11 for the
    ;; Australian user.
    ))

```

In summary, the calendar rendered currently accurately reflects the UTC date/time since that's the underlying model, but the user is likely to be confused unless they are also in UTC date/time. Once you change the value for :today to be local time, to avoid further confusion you'll likely end up wanting to change the underlying model to be a local Date. Or at least, that's the path I ended up going down.

**Comment by hipitihop** [comment]
@osbert Thanks again, and thanks for persisting. Further experiments show that `(goog.dateUtcDateTime.)` does not localise the date/time, perhaps my bad interpretation of the docs in the first place coupled with quirks of chrome debug tools printing localised time etc. Either way, I now fully understand how depending on your timezone and time of day `today` could be wrong. I've setup a branch where I'm experimenting with normalising `today` based on local date time but maintaining all datepicker underlying model & minimum & maximum models still utc based (mainly because of other already mentioned concerns) If my experiment proves successful, I will push the branch for further tests. Do you think this approach will be acceptable ?

BTW, I have found a way to test on Linux where I can simulate a timezone for a given browser session, which helps in testing.

**Comment by hipitihop** [comment]
@osbert I have now pushed the branch. https://github.com/Day8/re-com/tree/feature/today-fix<br>
I have tested it simulating New Zealnd timeszone as it passed midnight 30 min ago pushing their local time to the 13th and today now correctly highlighted, while our local GMT+1000 timezone still shows as 12th. Would you be please confirm this now solves the issue and we'll merge.

**Comment by osbert** [comment]
The other main reason to want to have a local date is because local
time midnight is likely the base offset for further calculations if
you also want to incorporate a time input field.

In the case where we use UTC as the model, if we select 2015-05-10, what we'll
get back is the following UTC date-time:

```clojure
_> (cljs-time.format/unparse (:rfc822 cljs-time.format/formatters) (cljs-time.core/date-time 2015 5 10))
_=> "Sun, 10 May 2015 00:00:00 Z"
```

Note that this represents UTC midnight, not local midnight. If I
simultaneously wish to ask the user for a time and then apply that
offset, I will need to first find midnight at local time anyways.

Let's say the user enters 730 in the corresponding time input widget. Apply an offset of 7h 30mins would take me to "Sun, 10 May 2015 07:30:00 Z" but the user actually intended for the date-time pair to result in "Sun, 10 May 2015 07:30:00 -07:00", or whatever their local time zone is.

**Comment by osbert** [comment]
I'd argue that it is significantly less error-prone and more accurate to say that when the user is entering a Date, they are intending to specify the equivalent of local time midnight.

Consider the following operations (assuming user offset -2):

1a) User specifies date (2015-05-10)
2a) Date is stored in model (2015-05-10 0000 UTC) / (2015-05-09 2200 -2)
3a) Date is rendered (2015-05-10 0000 UTC) / (2015-05-09 2200 -2)

In the general case, people will attempt to render to local date-time since they assume users don't care or will be confused by UTC.  But if you attempt to render this date by converting to local date it will show the wrong date. This is extremely counter-intuitive and confusing.

**Comment by hipitihop** [comment]
@osbert I apologise for my tardy response. We are going to have a look at what's involved to support either. e.g. in your use case you might provide a goog.date.DateTime as the :model in which case you would expect a goog.date.DateTime back, alternatively you might provide a goog.date.UtcDateTime Would that work for you ?

**Comment by osbert** [comment]
I think that's fine. I seem to recall from my brief experimentation that simply relaxing the constraint in goog-date? to check if it was a goog.Date.DateTime worked and was sufficient for my purposes. Thanks for the reply.

If you don't mind, please also let me know about your current use case and how you are able to stay in UTC end-to-end, I'm very curious if  there is an approach I'm missing. Thanks.

**Comment by hipitihop** [comment]
@osbert Apologies once again, we have only just managed to get back to reviewing outstanding re-com issues with the view to do a 0.6.0 release soon. Back in may I addressed the issue of the datepicker  incorrectly displaying today's date for users [please see this commit](https://github.com/Day8/re-com/commit/5ac6c424d88740a2b802fa6c51d0a971fa7d2314) Can you confirm this addresses the problem ? I realise this does not alter the underlying model which I agree with the benefit of hindsight could have been plain goog.date.Date We were driven by the defaults in cljs-time and our own use case which suited UTC. But afaict, really the only thing visible for users is the the today highlight (now fixed, pending your confirmation), and any consuming code (:on-change handler) can optionally convert the result back to local or any other timezone if needed. Let me know if you agree and we can close this.

**Comment by osbert** [comment]
Closing.
